heal/vlmd/validate/utils.py

```diff
@@ -0,0 +1,134 @@
+import charset_normalizer
+import jsonschema
+import pandas as pd
+from pathlib import Path
+
+from cdislogging import get_logger
+from heal.vlmd.config import CSV_SCHEMA, JSON_SCHEMA
+
+logger = get_logger("VLMD_UTILS", log_level="debug")
+
+
+def detect_file_encoding(file_path):
+    """
+    detects file encoding using charset_normalizer package
+    """
+
+    with open(file_path, "rb") as f:
+        data = f.read()
+        encoding_for_input = charset_normalizer.detect(data)
+
+    is_confident = encoding_for_input["confidence"] == 1
+    if not is_confident:
+        logger.warning("Be careful, the detected file encoding for:")
+        logger.warning(f"{file_path}")
+        logger.warning(r"has less than 100% confidence")
+
+    return encoding_for_input["encoding"]
+
+
+def read_delim(file_path, castdtype="string"):
+    """
+    reads in a tabular file (ie spreadsheet) after detecting
+    encoding and file extension without any type casting.
+
+    currently supports csv and tsv
+
+    defaults to not casting values (ie all columns are string dtypes)
+    and not parsing strings into NA values (eg "" is kept as "")
+    """
+    ext = Path(file_path).suffix
+    if ext == ".csv":
+        sep = ","
+    elif ext == ".tsv":
+        sep = "\t"
+    else:
+        raise ValueError("Delimited file must be csv or tsv")
+
+    encoding = detect_file_encoding(file_path)
+    file_encoding = pd.read_csv(
+        file_path, sep=sep, encoding=encoding, dtype=castdtype, keep_default_na=False
+    )
+
+    return file_encoding
+
+
+def get_schema(input_file: str, schema_type: str):
+    """
+    Get the schema for the specified schema_type.
+    Use the input_file suffix for "auto".
+
+    Args:
+        input_file (str): the path of the input VLMD file to be validated
+        schema_type (str): the type of the schema that can be validated against.
+            Allowed values for now are “csv”, “tsv”, “json” and “auto”. Defaults to “auto”
+    Returns:
+        schema
+    """
+
+    file_suffix = Path(input_file).suffix.replace(".", "")
+    if schema_type == "csv" or (schema_type == "auto" and file_suffix == "csv"):
+        schema = {"type": "array", "items": CSV_SCHEMA}
+        logger.debug("Validator will use CSV schema")
+        return schema
+    elif schema_type == "tsv" or (schema_type == "auto" and file_suffix == "tsv"):
+        schema = {"type": "array", "items": CSV_SCHEMA}
+        logger.debug("Validator will use CSV schema for TSV file")
+        return schema
+    elif schema_type == "json" or (schema_type == "auto" and file_suffix == "json"):
```

Comment:
nitpick: they don't necessarily need to be `elif` since the `if` block before them returns